pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -569,27 +568,46 @@ func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, s
 }
 
 // Delete deletes the component
-func (a Adapter) Delete(labels map[string]string) error {
-	spinner := log.Spinnerf("Deleting devfile component %s", a.ComponentName)
-	defer spinner.End(false)
+func (a Adapter) Delete(labels map[string]string, show bool) error {
 
-	componentExists, err := utils.ComponentExists(a.Client, a.ComponentName)
+	log.Infof("\nGathering information for component %s", a.ComponentName)
+	podSpinner := log.Spinner("Checking status for component")
+	defer podSpinner.End(false)
+
+	pod, err := a.Client.GetPodUsingComponentName(a.ComponentName)
 	if kerrors.IsForbidden(err) {
 		klog.V(4).Infof("Resource for %s forbidden", a.ComponentName)
 		// log the error if it failed to determine if the component exists due to insufficient RBACs
-		spinner.End(false)
+		podSpinner.End(false)
 		log.Warningf("%v", err)
 		return nil
+	} else if e, ok := err.(*kclient.NoPodFoundError); ok {
+		podSpinner.End(false)
+		log.Warningf("%v", e)
+		return nil
 	} else if err != nil {
 		return errors.Wrapf(err, "unable to determine if component %s exists", a.ComponentName)
 	}
 
-	if !componentExists {
-		spinner.End(false)
-		log.Warningf("Component %s does not exist", a.ComponentName)
-		return nil
+	podSpinner.End(true)
+
+	// if there are preStop events, execute them before deleting the deployment
+	preStopEvents := a.Devfile.Data.GetEvents().PreStop
+	if len(preStopEvents) > 0 {
+		if pod.Status.Phase != corev1.PodRunning {
```

Comment:
Rather than manually checking if the phase is `PodRunning`, can we just call `WaitAndGetPod` when we initally retrieve the pod and set the phase to wait for too `PodRunning`?

Reply:
It was a conscious decision to do this. A while ago, @mik-dass reviwed Stephanie's PR that I was also reviewing and had this same conversation https://github.com/openshift/odo/pull/3400#pullrequestreview-443650340

So I did this to be consistent and since @mik-dass is also reviewing this PR. Also, since we need the pod name, we already got the pod info in the beginning of `Delete()`, and agree with @mik-dass on the other PR that there is no need to do an API call again.

Reply:
Could we wrap it into a function then when we get the pod? We retrieve the pod and verify `pod.Status.Phase` with the requested phase?

Reply:
Don't really want to introduce another helper/wrapper func, there are already so many in pods.go https://github.com/openshift/odo/blob/master/pkg/kclient/pods.go#L154 and we're not being uniform in its usage everywhere. What is the wrapper going to achieve here, just return the pod name and status?

Reply:
That's fair.

My concern is that if we're not going to be using `WaitAndGetPod`, there's going to be loads of duplicated code in odo that pretty much does this (gets the pod, and then retrieves / checks the status).

Reply:
This case may not fit well for a helper function, since we're not checking the pod status until a bit after we get the pod, but in general, if we're frequently getting a pod and immediately checking its status, then that should be rolled up into a helper function.

Reply:
Hmm i see your point, but it will pretty much be the same thing:

currently:
```
pod, err := a.Client.GetPodUsingComponentName(a.ComponentName)
	if err != nil {
		return fmt.Errorf("error occurred while getting the pod: %w", err)
	}
	if pod.Status.Phase != corev1.PodRunning {
		return fmt.Errorf("pod for component %s is not running", a.ComponentName)
	}
```

Some callers are using `pod.Name` and all are checking for the status. So will the wrapper just be returning the pod{}, err, isPodRunning ?